src/webpub_manifest_parser/core/ast.py

```diff
@@ -0,0 +1,831 @@
+                frozenset(self.links),
+            )
+        )
+
+    def __repr__(self):
+        """Return a string representation of the object.
+
+        :return: String representation
+        :rtype: str
+        """
+        return u"<Contributor(name={0}, identifier={1}, sort_as={2}, role={3}, position={4}, links={5})>".format(
+            encode(self.name),
+            self.identifier,
+            self.sort_as,
+            self.role,
+            self.position,
+            self.links,
+        )
+
+
+class ContributorProperty(ParsableProperty):
+    """Property containing information about a contributor."""
+
+    PARSER = AnyOfParser(
+        [
+            StringParser(),
+            ArrayParser(AnyOfParser([StringParser(), TypeParser(Contributor)])),
+            TypeParser(Contributor),
+        ]
+    )
+
+
+class Subject(Node, PropertiesGrouping):
+    """Subject object."""
+
+    name = LocalizableStringProperty("name", required=True)
+    sort_as = StringProperty("sortAs", required=False)
+    code = StringProperty("code", required=False)
+    scheme = URIProperty("scheme", required=False)
+    links = ArrayOfLinksProperty("links", required=False)
+
+    def __hash__(self):
+        """Calculate the hash.
+
+        :return: Hash
+        :rtype: int
+        """
+        return hash(
+            (self.name, self.sort_as, self.code, self.scheme, frozenset(self.links))
+        )
+
+    def __repr__(self):
+        """Return a string representation of the object.
+
+        :return: String representation
+        :rtype: str
+        """
+        return (
+            u"<Subject(name={0}, sort_as={1}, code={2}, scheme={3}, links={4})>".format(
+                self.name, self.sort_as, self.code, self.scheme, self.links
+            )
+        )
+
+
+class SubjectProperty(ParsableProperty):
+    """Property containing information about a subject."""
+
+    PARSER = AnyOfParser(
+        [
+            StringParser(),
+            ArrayParser(AnyOfParser([StringParser(), TypeParser(Subject)])),
+            TypeParser(Subject),
+        ]
+    )
+
+
+class Owner(Node, PropertiesGrouping):
+    """Object containing information about the collection's owners."""
+
+    collection = TypeProperty("collection", required=False, nested_type=Contributor)
+    series = TypeProperty("series", required=False, nested_type=Contributor)
+
+    def __hash__(self):
+        """Calculate the hash.
+
+        :return: Hash
+        :rtype: int
+        """
+        return hash((self.collection, self.series))
+
+    def __repr__(self):
+        """Return a string representation of the object.
+
+        :return: String representation
+        :rtype: str
+        """
+        return u"<Owner(collection={0}, series={1})>".format(
+            self.collection, self.series
+        )
+
+
+class Metadata(Node):
+    """Dictionary containing manifest's metadata."""
+
+    identifier = URIProperty("identifier", required=False)
```

Comment:
Although the spec suggests that `identifier` is an URL, in practice it might not always be the case depending on the format.

Reply:
I was referring to the spec where it's defined as `uri`:
```
"identifier": {
      "type": "string",
      "format": "uri"
    },
```

Reply:
The JSON schema is used for "real" RWPM and OPDS 2, in this case it's true that `identifier` is an URL. But R2 shared models can be used to parse other formats so they are only loosely based on the JSON schema. For example EPUB and PDF can have identifiers that are not URL.

Another example is that RWPM requires one link with the rel `self`, but when parsing a local package (EPUB, PDF), we don't have any `self` link to parse so this requirement is lifted for the shared models.

Reply:
I see, in this case we could override this behaviour in child classes. For example:
```python
class EPUBMetadata(Metadata):
    identifier = StringProperty("identifier", required=False)

Metadata.extensions = (EPUBMetadata, )
```

It will force the parser to instantiate `EPUBMetadata` instead of `Metadata` which means that it will be expecting an identifier to be a string, not URI. Does it make sense?

Reply:
I think so, on mobile we have only a single `Metadata` class so we're not in this situation, but maybe it's a proper workaround.